### PR TITLE
Added Dynamic Power and Accuracy in Move Descriptions

### PR DESCRIPTION
**Pull request by Flash1Lucky**
Adds a configuration option to display the actual base power and accuracy of moves in battle move descriptions, factoring in ability, stat, and item modifiers. Additionally, it introduces individual configuration options to control which factors (abilities, stats, items, weather, terrain, etc.) affect these values. This extends the existing Move Descriptions feature introduced in PR #4152.

Screenshot 2024-09-05 180225

## **Known Issues**

- Earthquake/Magnitude in Grassy Terrain while opponent is semi-invulnerable and on the ground isn't displayed right currently.

## **People who collaborated with me in this PR**
Code based on Move Descriptions feature from @PCG06 which was based on the implementation from Elite Redux.
Improved and reviewed by @AlexOn1ine.

## **Discord contact info**
flash1lucky

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
Calcs are stored in simulatedDamage so instead of recalc'ing everything which will waste a lot of cycles you should fetch the damage from there.

**Review by AlexOn1ine — CHANGES_REQUESTED** [review]
Calcs are stored in simulatedDamage so instead of recalc'ing everything which will waste a lot of cycles you should fetch the damage from there.

Also the config should be off by default

**Comment by AlexOn1ine** [comment]
In battle.h there is a struct called `AiLogicData`. There is already a field preset called moveAccuracy. Can you add this additional field to the struct:
```
u8 moveBasePower[MAX_BATTLERS_COUNT][MAX_BATTLERS_COUNT][MAX_MON_MOVES];
```
Then in function `DoMoveDamageCalcVars` can you assign `gBattleMovePower` to that field?
then you can remove all the calcs you are doing and use both fields instead .

The problem with the current implementation is that it will recalc everything each time you open the move description but we already calc it at the beginning of the turn so there is no reason to do everything again

**Comment by kittenchilly** [comment]
Should be off by default anyways I think, no official game gives ever you this info

**Comment by Flash1Lucky** [comment]
> Should be off by default anyways I think, no official game gives ever you this info

Yeah, that's the plan

File: src/battle_util.c [hunk 1]
```diff
@@ -9463,6 +9463,46 @@ static inline u32 CalcMoveBasePowerAfterModifiers(u32 move, u32 battlerAtk, u32
...
+
+            if (move == MOVE_NONE 
+            || move == 0xFFFF
+            || move == MOVE_KNOCK_OFF
+            || move == MOVE_BRINE
+            || move == MOVE_EARTHQUAKE
+            || move == MOVE_MAGNITUDE)
+                continue;
+
+
```
Comment (AlexOn1ine) [inline src/battle_util.c:9493]:
```suggestion
            if (move == MOVE_NONE 
            ||  move == 0xFFFF
            ||  move == MOVE_KNOCK_OFF
            ||  move == MOVE_BRINE
            ||  move == MOVE_EARTHQUAKE
            ||  move == MOVE_MAGNITUDE)
                continue;

```

File: src/battle_util.c [hunk 2]
```diff
@@ -9463,6 +9463,46 @@ static inline u32 CalcMoveBasePowerAfterModifiers(u32 move, u32 battlerAtk, u32
     return uq4_12_multiply_by_int_half_down(modifier, basePower);
 }
 
+void SetMoveUIDataForTurn(struct MoveUIData *moveUIData)
+{
+    u32 battlerAtk, battlerDef, i, movePower, moveAccuracy, atkAbility, holdEffectAtk;
+    u32 weather = gBattleWeather;
```
Comment (AlexOn1ine) [inline src/battle_util.c:9469]: you can just remove this var and pass `gBattlerWeather` to the funct

File: src/battle_main.c [hunk 3]
```diff
@@ -3744,6 +3744,9 @@ static void DoBattleIntro(void)
             gBattleStruct->overworldWeatherDone = FALSE;
             SetAiLogicDataForTurn(AI_DATA); // get assumed abilities, hold effects, etc of all battlers
             Ai_InitPartyStruct(); // Save mons party counts, and first 2/4 mons on the battlefield.
+            #if B_DYNAMIC_MOVE_DESCRIPTIONS == TRUE
+                SetMoveUIDataForTurn(MOVE_UI_DATA);
+            #endif
```
Comment (AlexOn1ine) [inline src/battle_main.c:3749]: what I would do is move all those call to `Ai_InitPartyStruct` since this one will be always code it saves some lines

**Comment by AlexOn1ine** [comment]
Flash, I've been thinking about and I think it is better when the calculations are   computed on the fly when the move description is accessed. Reason for is that the calcs use a lot of cycles and it is better

File: src/battle_controller_player.c [hunk 4]
```diff
@@ -1764,11 +1764,38 @@ static void MoveSelectionDisplayMoveType(u32 battler)
...
+    u32 atkAbility = GetBattlerAbility(battlerAtk);
+    u32 holdEffectAtk = GetBattlerHoldEffect(battlerAtk, TRUE);
     u8 cat = gMovesInfo[move].category;
 
+    if (B_DYNAMIC_MOVE_DESCRIPTIONS)
+    {
+        u32 moveEffect = gMovesInfo[move].effect;
+
+        if (move == MOVE_NONE || move == 0xFFFF || moveEffect == EFFECT_KNOCK_OFF || moveEffect == EFFECT_BRINE)
+            return;
```
Comment (AlexOn1ine) [inline src/battle_controller_player.c:1782]: this part should probably be included the previous if statement so it branches out to the else statement

**Comment by pkmnsnfrn** [comment]
I'm happy to test this on stream this weekend. @AlexOn1ine are there specific things and cases you'd like me to test and make sure are covered?

**Pull request closed without merging**